tracer/src/Datadog.Trace/Logging/DirectSubmission/Sink/PeriodicBatching/BatchingSink.cs

```diff
@@ -0,0 +1,223 @@
+        /// </summary>
+        /// <param name="events">The events to emit.</param>
+        /// <returns><c>true</c> if the batch was emitted successfully. <c>false</c> if there was an error</returns>
+        protected abstract Task<bool> EmitBatch(Queue<DatadogLogEvent> events);
+
+        private async Task FlushBuffersTaskLoopAsync()
+        {
+            while (true)
+            {
+                if (_processExit.Task.IsCompleted)
+                {
+                    _log.Debug("Terminating Log submission loop");
+                    if (_processExit.Task.Result)
+                    {
+                        var maxShutDownDelay = Task.Delay(20_000);
+                        var finalFlushTask = FlushLogs();
+                        var completed = await Task.WhenAny(finalFlushTask, maxShutDownDelay).ConfigureAwait(false);
+
+                        if (completed != finalFlushTask)
+                        {
+                            _log.Warning("Could not finish flushing all logs before process end");
+                        }
+                    }
+
+                    return;
+                }
+
+                var circuitStatus = await FlushLogs().ConfigureAwait(false);
+
+                await HandleCircuitStatus(circuitStatus).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<CircuitStatus> FlushLogs()
+        {
+            try
+            {
+                var status = CircuitStatus.Closed;
+                var haveMultipleBatchesToSend = false;
+                do
+                {
+                    while (_waitingBatch.Count < _batchSizeLimit &&
+                           _queue.TryDequeue(out var next))
+                    {
+                        _waitingBatch.Enqueue(next);
+                    }
+
+                    if (_waitingBatch.Count == 0)
+                    {
+                        // If the first batch was full, then use that status.
+                        // If this is the first batch in this loop, then there were no logs to send
+                        // So can't say anything about the status of the API.
+                        return haveMultipleBatchesToSend ? status : _circuitBreaker.MarkSkipped();
+                    }
+
+                    var success = await EmitBatch(_waitingBatch).ConfigureAwait(false);
+                    if (success)
+                    {
+                        status = _circuitBreaker.MarkSuccess();
+                        haveMultipleBatchesToSend = _waitingBatch.Count >= _batchSizeLimit;
+                        if (haveMultipleBatchesToSend)
+                        {
+                            _waitingBatch.Clear();
+                        }
+                    }
+                    else
+                    {
+                        return _circuitBreaker.MarkFailure();
+                    }
+                }
+                while (haveMultipleBatchesToSend);
```

Comment:
I'd argue that's a feature rather than a bug - if we need to constantly send logs to keep up with the app, then the alternative would effectively be a memory leak